libfreerdp/codec/progressive.c

```diff
@@ -2308,10 +2318,253 @@ INT32 progressive_decompress(PROGRESSIVE_CONTEXT* progressive, const BYTE* pSrcD
+	Stream_Write_UINT8(s, 0);              /* flags (1 byte) */
+	Stream_Write_UINT16(s, msg->numTiles); /* numTiles (2 bytes) */
+	Stream_Write_UINT32(s, tilesDataSize); /* tilesDataSize (4 bytes) */
+
+	for (i = 0; i < msg->numRects; i++)
+	{
+		/* TS_RFX_RECT */
+		Stream_Write_UINT16(s, msg->rects[i].x);      /* x (2 bytes) */
+		Stream_Write_UINT16(s, msg->rects[i].y);      /* y (2 bytes) */
+		Stream_Write_UINT16(s, msg->rects[i].width);  /* width (2 bytes) */
+		Stream_Write_UINT16(s, msg->rects[i].height); /* height (2 bytes) */
+	}
+
+	/**
+	 * Note: The RFX_COMPONENT_CODEC_QUANT structure differs from the
+	 * TS_RFX_CODEC_QUANT ([MS-RDPRFX] section 2.2.2.1.5) structure with respect
+	 * to the order of the bands.
+	 *             0    1    2   3     4    5    6    7    8    9
+	 * RDPRFX:   LL3, LH3, HL3, HH3, LH2, HL2, HH2, LH1, HL1, HH1
+	 * RDPEGFX:  LL3, HL3, LH3, HH3, HL2, LH2, HH2, HL1, LH1, HH1
+	 */
+	for (i = 0, qv = msg->quantVals; i < msg->numQuant; i++, qv += 10)
+	{
+		/* RFX_COMPONENT_CODEC_QUANT */
+		Stream_Write_UINT8(s, qv[0] + (qv[2] << 4)); /* LL3 (4-bit), HL3 (4-bit) */
+		Stream_Write_UINT8(s, qv[1] + (qv[3] << 4)); /* LH3 (4-bit), HH3 (4-bit) */
+		Stream_Write_UINT8(s, qv[5] + (qv[4] << 4)); /* HL2 (4-bit), LH2 (4-bit) */
+		Stream_Write_UINT8(s, qv[6] + (qv[8] << 4)); /* HH2 (4-bit), HL1 (4-bit) */
+		Stream_Write_UINT8(s, qv[7] + (qv[9] << 4)); /* LH1 (4-bit), HH1 (4-bit) */
+	}
+
+	for (i = 0; i < msg->numTiles; i++)
+	{
+		/* RFX_PROGRESSIVE_TILE_SIMPLE */
+		tile = msg->tiles[i];
+		blockLen = 22 + tile->YLen + tile->CbLen + tile->CrLen;
+		Stream_Write_UINT16(s, 0xCCC5);             /* blockType (2 bytes) */
+		Stream_Write_UINT32(s, blockLen);           /* blockLen (4 bytes) */
+		Stream_Write_UINT8(s, tile->quantIdxY);     /* quantIdxY (1 byte) */
+		Stream_Write_UINT8(s, tile->quantIdxCb);    /* quantIdxCb (1 byte) */
+		Stream_Write_UINT8(s, tile->quantIdxCr);    /* quantIdxCr (1 byte) */
+		Stream_Write_UINT16(s, tile->xIdx);         /* xIdx (2 bytes) */
+		Stream_Write_UINT16(s, tile->yIdx);         /* yIdx (2 bytes) */
+		Stream_Write_UINT8(s, 0);                   /* flags (1 byte) */
+		Stream_Write_UINT16(s, tile->YLen);         /* YLen (2 bytes) */
+		Stream_Write_UINT16(s, tile->CbLen);        /* CbLen (2 bytes) */
+		Stream_Write_UINT16(s, tile->CrLen);        /* CrLen (2 bytes) */
+		Stream_Write_UINT16(s, 0);                  /* tailLen (2 bytes) */
+		Stream_Write(s, tile->YData, tile->YLen);   /* YData */
+		Stream_Write(s, tile->CbData, tile->CbLen); /* CbData */
+		Stream_Write(s, tile->CrData, tile->CrLen); /* CrData */
+	}
+
+	/* RFX_PROGRESSIVE_FRAME_END */
+	blockLen = 6;
+	if (!Stream_EnsureRemainingCapacity(s, blockLen))
+	{
+		return FALSE;
+	}
+	Stream_Write_UINT16(s, 0xCCC2);   /* blockType (2 bytes) */
+	Stream_Write_UINT32(s, blockLen); /* blockLen (4 bytes) */
+
+	return TRUE;
+}
+
 int progressive_compress(PROGRESSIVE_CONTEXT* progressive, const BYTE* pSrcData, UINT32 SrcSize,
```

Comment:
Maybe this could be renamed to `progressive_compress_simple`? This would leave space for a potential `progressive_compress_first` and `progressive_compress_upgrade` function for progressive encoding.

Reply:
plan is to let the compression routine decide that at a later point in time according to input data, so no manual switch is required.

Reply:
Ah ok. That sounds great.